src/content/docs/concepts/widgets.md

```diff
@@ -0,0 +1,308 @@
+
+```rust
+pub trait Widget {
+    fn render(self, area: Rect, buf: &mut Buffer);
+}
+```
+
+### StatefulWidget
+
+The [`StatefulWidget`] trait is similar to the [`Widget`] trait, but also includes a state that can
+be managed and updated during rendering.
+
+```rust
+pub trait StatefulWidget {
+    type State;
+    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State);
+}
+```
+
+### WidgetRef and StatefulWidgetRef
+
+The [`WidgetRef`] trait allows for rendering a Widget by reference instead of consuming the widget,
+which can be useful for storing and rendering individual or collections of widgets.
+
+The [`StatefulWidgetRef`] trait is similar to the [`WidgetRef`] trait, but also includes a state
+that can be managed and updated during rendering.
+
+These two traits were introduced in Ratatui 0.26.0 to help avoid a shortcoming that meant that
+widgets were always consumed on rendering while not breaking all code that has previously been built
+with that assumption. These two widgets are currently marked as unstable and gated behind the
+`unstable-widget-ref` feature flag.
+
+```rust
+pub trait WidgetRef {
+    fn render_ref(&self, area: Rect, buf: &mut Buffer);
+}
+
+pub trait StatefulWidgetRef {
+    type State;
+    fn render(&self, area: Rect, buf: &mut Buffer, state: &mut Self::State);
+}
+```
+
+# Using Widgets
+
+In Ratatui, widgets are used to create and manage the layout and style of the terminal interface.
+widgets can be combined and nested to create complex UIs, and can be easily customized to suit the
+needs of your application.
+
+To use widgets in your application, you will typically use the [`Frame`] type, which has two methods
+for rendering widgets: [`render_widget`] and [`render_stateful_widget`] (and the corresponding
+`_ref` methods). These methods are the entry points for an application to draw widgets and are
+usually called from the closure that is passed to the [`Terminal::draw`] method.
+
+Here's an example of using the `render_widget` method to draw a Widget:
+
+```rust
+terminal.draw(|frame| {
+    frame.render_widget(some_widget, frame.size());
+});
+```
+
+And here's an example of using the `render_stateful_widget` method to draw a StatefulWidget:
+
+```rust
+terminal.draw(|frame| {
+    frame.render_stateful_widget(some_stateful_widget, frame.size(), &mut some_state);
+});
+```
+
+These methods internally call the `render` function on the [`Widget`] or [`StatefulWidget`] trait,
+which will then call the `render` function on the specific Widget or StatefulWidget that you have
+implemented.
+
+A common compositional pattern is to only have a single Root widget (perhaps your application or
+similar) that is passed to `Frame::render_widget()` and then within that and other widgets, you call
+the render methods directly passing in the area which you want to render the widgets to.
+
+```rust
+MyHeaderWidget::new("Header text").render(Rect::new(0, 0, area.width, 1), buf);
+```
```

Comment:
I think I understand what you are trying to say here. What do you think about expanding the code example here more? e.g.

```rust
fn main() {
    let terminal = Terminal::new();
    let app = App::new();
    while app.should_quit {
        terminal.draw(|frame| {
            frame.render_widget(&app)
        })
    }
}

impl Widget for App {
    fn render(area: Rect, buf: &mut Buffer) {
        MyHeaderWidget::new("Header text")
            .render(Rect::new(0, 0, area.width, 1), buf);
    }
}
```

Reply:
Sure - go for it - though you need to make that &Root (and I'd suggest making it App).